AK: Address is_power_of rounding issues when base != 2

abuneri opened the pull request:
### Context
I'm currently working on adding MSVC support for AK to make it a fully cross platform library, with 565 tests and 10 benchmarks passing at the moment.

![msvc_AK_tests](https://github.com/SerenityOS/serenity/assets/11325341/62eb3fa6-79f3-4b78-8e5e-b4b5c11d381d)

Besides a lot of ifdef's to replace GCC/Clang-specific compiler built-ins with MSVC equivalents, so far I have not had to change/touch the existing code for any of the tests I have gotten to pass on MSVC.

### Problem
On MSVC in `TestIntegerMath.cpp` the `is_power_of` test case was getting a lot of failues with `prime != 2`.
![msvc_intmath_test_fail](https://github.com/SerenityOS/serenity/assets/11325341/01697724-dde9-4213-8081-0c5477da572d)

This new implementation makes all the tests pass on MSVC. Since that code had a FIXME comment I decided to submit my fix to upstream.

### Results

#### MSVC
![msvc_intmath_test](https://github.com/SerenityOS/serenity/assets/11325341/336aa023-381a-45b3-a39e-ac5a2900925c)

#### GCC
![gcc_intmath_test](https://github.com/SerenityOS/serenity/assets/11325341/1d2c0dfb-342e-4994-a87b-31d897fa31cc)

DanShaders commented:
Welcome to the project!

Just out of curiosity: how have you worked around the absence of statement expressions in MSVC? We make heavy use of TRY/MUST macros in AK and replacing every single instance of them (even just in AK) seems monumental. AFAIC, this was also the reason we decided to not even try targeting MSVC.

That said, regardless of the original motivation, the change itself looks good to me. The only problem I see is the commit title/description. I think we try to be a bit more specific in them. So, instead of "addressing vague issues somewhere", something like
```
AK: Replace FP math in `is_power_of` with a purely integral algorithm

The previous naive approach was causing test failures because of
rounding issues in some exotic environments.
```
would be a lot better.

abuneri commented:
> how have you worked around the absence of statement expressions in MSVC?

Yeah, the use of statement expressions is very unfortunate wrt MSVC compatibility, really wish Microsoft would implement it too. 

For the `MUST` macro there is a workaround via a lambda that just works; however, for the `TRY` macro any attempt at an equivalent just using the lambda workaround was met with the obvious problem that once a return type is deduced you can't attempt to return a different type.

Serenity's usage of the `TRY` macro seems to equate to returning `T` in the happy path but propagating `ErrorType` in the sad path, where eventually the root caller handles the `ErrorType` in some way (or a caller somewhere else in the stack handles it before potentially propagating their own `ErrorOr` that eventually makes it to the root).

This is essentially the pattern of exceptions, so after spending quite awhile trying to find something that could work, I adjusted `ErrorOr` to have a helper method (MSVC only) that throws an exception while capturing the `ErrorType` instance in the exception. Then in the `TRY` macro the `ErrorOr` instance calls that helper (using the same lambda pattern that allowed `MUST` to work) so we no longer have the multi return type issue as we're throwing a wrapper around `ErrorType` instead of returning it explicitly.

Then at the call sites that want to handle `ErrorOr` (I've come across a few in some tests that expect errors, but nothing in AK itself I've added support for explicitly handles `ErrorOr`s), I created helper macros for the methods like `is_error`, `error`, `release_error` etc that on MSVC wrap that call-site in a try-catch and then return the equivalent of those methods based on the `ErrorType` instance that the exception has.

This is the only way I can find to allow all the `TRY` code to be unchanged, so while it's not pretty, this is the only way that things like Ladybird could ever be run natively on Windows unless all the `TRY` error handling in the code base is re-written to not use expression statements or MSVC adds support for statement expressions.

DanShaders commented:
> this is the only way that things like Ladybird could ever be run natively on Windows

That's not true. I think the consensus is to use clang with `x86_64-pc-windows-msvc` target. This is how jakt works on Windows, for example. There's also a stale PR (#21739) that made LibJS work with this approach.

abuneri commented:
Ah true, the clang-cl approach probably involves less work. But my goal was to get AK working without Clang/GCC at all. I also have no plan of wanting my stuff to be upstreamed, this is just a personal challenge I wanted to try

abuneri commented:
@DanShaders FYI the fix `is_power_of` fix moved to https://github.com/SerenityOS/serenity/pull/24218 with more explicit testing for some edge cases
